component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java

```diff
@@ -291,22 +331,33 @@ protected void processAuthenticationResponse(HttpServletRequest request,
             throws AuthenticationFailedException {
 
         String token = request.getParameter(TOTPAuthenticatorConstants.TOKEN);
-        String username = context.getProperty("username").toString();
-        validateAccountLockStatusForLocalUser(context, username);
+        AuthenticatedUser authenticatedUser = getAuthenticatedUser(context);
+        String username = authenticatedUser.toFullQualifiedUsername();
+        String tenantDomain = authenticatedUser.getTenantDomain();
+        String userId = (String) context.getProperty(TOTPAuthenticatorConstants.FEDERATED_USER_ID);
+        validateAccountLockStatusForLocalUser(context, authenticatedUser);
         if (StringUtils.isBlank(token)) {
             handleTotpVerificationFail(context);
             throw new AuthenticationFailedException("Empty TOTP in the request. Authentication Failed for user: " +
                     username);
         }
-        checkTotpEnabled(context, username);
         try {
             int tokenValue = Integer.parseInt(token);
-            if (!isValidTokenLocalUser(tokenValue, username, context)) {
-                handleTotpVerificationFail(context);
-                throw new AuthenticationFailedException("Invalid Token. Authentication failed, user :  " + username);
+            if (TOTPUtil.isLocalUser(context)) {
+                checkTotpEnabled(context, username);
+                if (!isValidTokenLocalUser(tokenValue, username, context)) {
+                    handleTotpVerificationFail(context);
+                    throw new AuthenticationFailedException("Invalid Token. Authentication failed, user :  " + username);
```

Comment:
Shall we update the error message with more context, something similar to below:

throw new AuthenticationFailedException("Invalid Token. Authentication failed for local user :  " + username in the tenant domain : + tenantDomain);

Reply:
Fixed https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/94/commits/01e2f562f9e60129c3810e75f2cb8f16020632bc